### PR TITLE
feat: Prevent driver from stringifying binary responses

### LINKED ISSUE #16722: cy.intercept doesn't handle binary data correctly
### Current behavior
1. When intercepting a request that returns binary data (for example `application/octet-stream`) the response is corrupted by `cy.intercept`.
2. When trying to stub a request with an `ArrayBuffer` it is interpreted as JSON.

### Desired behavior
1. Intercepting a request does not alter the response if it is not stubbed.
2. It is possible to supply an `ArrayBuffer` to an intercept stub.

### Test code to reproduce
Reproduction code is here: https://github.com/simfeld/cypress-intercept-binary. There are three tests:

1. no intercept: This test succeeds, the response is loaded correctly from the server.
2. intercept without stub: This test fails. The request is intercepted but not stubbed. The response that arrives in the browser is `{}` or `[123, 125]` when interpreted as `ArrayBuffer`, instead of the response that the server delivers.
3. intercept with stub: This test fails. The request is intercepted and a stub is returned. The response that arrives in the browser is `{}` or `[123, 125]` when interpreted as `ArrayBuffer`, instead of the stub supplied to the intercept.

### Versions

Latest version as of filing this issue (`7.4.0`)

### Related issues

This issue is possibly related: https://github.com/cypress-io/cypress/issues/15038

### PR DESCRIPTION
**Pull request by simfeld**
- Fixes #16722

### User facing changelog

- (fix) `cy.intercept()` does not corrupt binary responses
- (feat) You can now pass an `ArrayBuffer` as request / response body with `cy.intercept()`

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue or this PR been tagged with a release in ZenHub? 
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/driver/src/cy/net-stubbing/events/response.ts
Comment (simfeld) [inline packages/driver/src/cy/net-stubbing/events/response.ts:107]: This is the main change. `_.isObject` returns true for many things including `ArrayBuffer`. The response body should only be stringified here if it was previously parsed, i.e. `hasJsonContentType` evaluated to true.

File: packages/net-stubbing/lib/server/middleware/response.ts
Comment (simfeld) [inline packages/net-stubbing/lib/server/middleware/response.ts:74]: `_.merge` [converts Buffers to Arrays](https://github.com/lodash/lodash/issues/2964), so we use `_.mergeWith` with a customizer that preserves the source buffer.

**Comment by jennifer-shehane** [comment]
Hey @simfeld, I notice this PR is still in draft. Do you have more work planned? We'd like tests to cover any fixes. Thanks!

**Comment by simfeld** [comment]
@jennifer-shehane I put it in draft since I haven't added tests. If you can give me some guidance where a test for this kind of fix should live (my best guess is [here](https://github.com/cypress-io/cypress/blob/develop/packages/driver/cypress/integration/commands/net_stubbing_spec.ts)) I can give it a shot. Otherwise if your team has capacity to work on this issue based on my analysis I'd also be happy with that.

File: packages/net-stubbing/lib/server/middleware/response.ts
Comment (flotwig) [reply packages/net-stubbing/lib/server/middleware/response.ts:74]: the same bug on the request-side was recently fixed like so: https://github.com/cypress-io/cypress/blob/9333bb61e5c88c7ce7ebca9b2351d63170d7325f/packages/net-stubbing/lib/server/middleware/request.ts#L145-L156

could you please update request.ts to use your implementation (which looks cleaner) and import it from somewhere so it can be shared b/t response.ts and request.ts?

File: packages/driver/src/cy/net-stubbing/events/utils.ts
Comment (flotwig) [inline packages/driver/src/cy/net-stubbing/events/utils.ts:22]:
```suggestion
export function stringifyJsonBody (message: CyHttpMessages.BaseMessage) {
```

to parallel `parseJsonBody`

**Review by flotwig — CHANGES_REQUESTED** [review]
Makes sense, just some minor changes.

Also, please add a test in `net_stubbing_spec` to ensure there's not a regression in the future with regards to this bug. Most likely in this section: https://github.com/cypress-io/cypress/blob/9333bb61e5c88c7ce7ebca9b2351d63170d7325f/packages/driver/cypress/integration/commands/net_stubbing_spec.ts#L2327-L2387

File: packages/driver/src/cy/net-stubbing/events/utils.ts
Comment (simfeld) [reply packages/driver/src/cy/net-stubbing/events/utils.ts]: 979be70

File: packages/net-stubbing/lib/server/middleware/response.ts
Comment (simfeld) [reply packages/net-stubbing/lib/server/middleware/response.ts]: 31ee892 suggestions for a better name are highly welcome :-P

File: packages/driver/cypress/integration/commands/net_stubbing_spec.ts
Comment (simfeld) [inline packages/driver/cypress/integration/commands/net_stubbing_spec.ts:2372]: @flotwig can you enlighten me about the `@see` annotation? I see it used in many places. I have read https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#adding-links-within-code but it doesn't mention what this does/means.

File: packages/driver/cypress/integration/commands/net_stubbing_spec.ts
Comment (simfeld) [inline packages/driver/cypress/integration/commands/net_stubbing_spec.ts:2391]: Asserting on the response body from the wait is not enough (that also works without the fix). We need to get the result in the browser and display it (without the fix we get "123, 125").

File: packages/driver/src/cy/net-stubbing/events/response.ts
Comment (simfeld) [inline packages/driver/src/cy/net-stubbing/events/response.ts:24]: We need to remember whether the body was parsed because the intercept could potentially remove the "content-type: application/json" header which would result in the body being parsed but not stringified (if `stringifyJsonBody` would rely on this header as in my initial implementation). This was exhibited by the "can delete a response header" test in net-stubbing.

**Comment by simfeld** [comment]
I added a change allowing to provide `ArrayBuffer`s as body in `cy.intercept` (228e666) ad corresponding tests (d547e4d). This achieves the desired behaviour no. 2 in #16722. I can also make a separate PR if you wish, but these changes only work on top of the previous ones in this PR.

File: packages/driver/src/cy/net-stubbing/events/response.ts
Comment (simfeld) [inline packages/driver/src/cy/net-stubbing/events/response.ts:103]: @flotwig this is a nasty one. The explanation I can come up with is this: The socket.io parser [uses](https://github.com/socketio/socket.io-parser/blob/af1b23ca85fb64a0d7a050abd0362c5f632ce429/lib/is-binary.ts#L27) `instanceof ArrayBuffer` to check whether a part of the object to be encoded is an ArrayBuffer. Now the place where the ArrayBuffer coming from the real server response and the place where a stubbed ArrayBuffer response are constructed are not in the same scope and thus the prototype of the respective ArrayBuffers are not the same, i.e. `instanceof` returns `false` even though the object provided is an ArrayBuffer but from a different context (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof#instanceof_and_multiple_context_e.g._frames_or_windows). So if you just use merge here you get a very puzzling behaviour:

- if you intercept a binary response but do not touch the body, everything works as expected (ArrayBuffer is detected as binary from socket.io parser and encoded correspondingly)
- if you intercept a binary response and replace the body by another ArrayBuffer you construct inside of your e2e test, it is encoded as text and the Cypress server receives `{}` instead of the encoded binary content.

The possible solutions I see are:

- Create a new ArrayBuffer and deep copy the binary values (implemented in [utils.ts](https://github.com/cypress-io/cypress/pull/16755/files#diff-b0318835d7e6e6b3346d8b261b8647d0848c58591e3242a87af81483cd52380b)). I'm not 100% sure though if the copying via `Uint8Array` works for all scenarios.
- Overwrite the prototype with [Object.setPrototypeOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf) (also implemented in [utils.ts](https://github.com/cypress-io/cypress/pull/16755/files#diff-b0318835d7e6e6b3346d8b261b8647d0848c58591e3242a87af81483cd52380b))
- Write a [custom parser](https://socket.io/docs/v3/custom-parser/) for socket.io that for example uses `_.isArrayBuffer` instead of `instanceof`.

Please let me know what you prefer to do here (I tend to go for the second option) so that I can adapt my PR accordingly.

File: packages/driver/cypress/integration/commands/net_stubbing_spec.ts
Comment (jennifer-shehane) [reply packages/driver/cypress/integration/commands/net_stubbing_spec.ts:2418]: `@see` is part of jsdoc, you can read about the see tag here: https://jsdoc.app/tags-see.html

In general we want a link in the comments to the relevant GitHub issue that the test or code is addressing for our team's future context.

File: packages/driver/src/cy/net-stubbing/events/response.ts
Comment (flotwig) [reply packages/driver/src/cy/net-stubbing/events/response.ts:103]: @simfeld Cypress already patches `socket.io-parser` to use `CircularJSON` to allow us to pass circular objects through `socket.io`: https://github.com/cypress-io/cypress/blob/77880f1a0f0cecaee92e1f271cede452cc133287/packages/socket/patches/socket.io-parser+4.0.4.patch#L1-L162

It would be good to extend the patch so that we can have consistent `socket.io` behavior throughout the codebase. There are tests for the patched parser in `packages/socket/test/socket_spec.js`. The patch is applied with the `patch-package` file, so to update it, make changes in `packages/socket/node_modules/socket.io-parser`, and run `yarn patch-package socket.io-parser` to update the diff.

Go with whatever makes the implementation the easiest to understand - from my understanding of your comment, this would probably be updating the patch, since it would remove the need for `mergeWithArrayBuffer`?

**Comment by flotwig** [comment]
@simfeld this is good to merge once https://github.com/cypress-io/cypress/pull/16755#discussion_r650225882 is finalized

File: packages/driver/src/cy/net-stubbing/events/response.ts
Comment (simfeld) [reply packages/driver/src/cy/net-stubbing/events/response.ts]: @flotwig thanks for the hint, that's clearly a cleaner solution. See 67e00da. Note that both my repro tests (https://github.com/simfeld/cypress-intercept-binary) and the tests I added in this PR also pass with the `engio.io-parser` patch _only_. I'm not quite sure in what scenarios `socket.io-parser` is used, but I included the same fix there anyway.

**Review by flotwig — APPROVED** [review]
LGTM, thanks for opening a PR @simfeld

**Comment by simfeld** [comment]
@flotwig I guess there are some user facing documentation updates required? Are you taking care of that?

**Comment by flotwig** [comment]
@simfeld If you could open a PR, it would be appreciated, this is the docs repo: https://github.com/cypress-io/cypress-documentation

**Comment by simfeld** [comment]
> @simfeld If you could open a PR, it would be appreciated, this is the docs repo: https://github.com/cypress-io/cypress-documentation

cypress-io/cypress-documentation#3984 @flotwig I think that should be sufficient.

**Comment by simfeld** [comment]
@flotwig any further action required on my side to get this merged? Or are you just waiting for a second reviewer from your team?

**Comment by flotwig** [comment]
> @flotwig any further action required on my side to get this merged? Or are you just waiting for a second reviewer from your team?

Nothing's required from your side, I am hitting some flake with the tests I think. Once it's passing it will be merged.

**Pull request merged**
